modal/image.py

```diff
@@ -1299,6 +1299,68 @@ def build_dockerfile(version: ImageBuilderVersion) -> DockerfileSpec:
             gpu_config=parse_gpu_config(gpu),
         )
 
+    def uv_sync(
+        self,
+        lockfile: str = "./uv.lock",  # Path to lockfile
+        # Path to pyproject.toml. If not provided, then use file in the same directory as the lockfile
+        pyproject_toml: Optional[str] = None,
+        *,
+        force_build: bool = False,  # Ignore cached builds, similar to 'docker build --no-cache'
+        secrets: Sequence[_Secret] = [],
+        gpu: GPU_T = None,
+    ) -> "_Image":
+        """Install dependencies from a local `uv.lock` file using `uv sync`.
+
+        Note that only dependencies are installed. Please use `add_local_python_source` to include local python source
+        files. Please ensure that `modal` is a dependency by running `uv add modal`.
+        """
+
+        def build_dockerfile(version: ImageBuilderVersion) -> DockerfileSpec:
+            lockfile_ = os.path.expanduser(lockfile)
+            if pyproject_toml is None:
+                parent_dir = os.path.dirname(lockfile_)
+                pyproject_toml_ = os.path.join(parent_dir, "pyproject.toml")
+                if not os.path.exists(pyproject_toml_):
+                    msg = "pyproject.toml does not exist in the same directory as the uv.lock file"
+                    raise InvalidError(msg)
+            else:
+                pyproject_toml_ = os.path.expanduser(lockfile)
+
+            uv_root = "/.uv"
+            context_files = {"/.uv.lock": lockfile_, "/.pyproject.toml": pyproject_toml_}
+
+            uv_sync_args = [
+                f"--project={uv_root}",
+                "--frozen",  # Do not update `uv.lock`
+                "--no-install-workspace",  # Do not install the root project or any "uv workspace"
+                "--no-cache",  # Cache is not persisted, so we disable it
+                "--no-managed-python",  # Use the system python interpreter to create venv
+                "--compile-bytecode",
+            ]
+            uv_sync_args_joined = " ".join(uv_sync_args)
+
+            # TODO Make user configurable through a kwargs
+            UV_VERSION = "0.7.11"
+
+            commands = [
+                "FROM base",
+                f"RUN python -m pip install uv=={UV_VERSION}",
+                f"COPY /.pyproject.toml {uv_root}/pyproject.toml",
+                f"COPY /.uv.lock {uv_root}/uv.lock",
```

Comment:
Does the `uv.lock` file get modified outside of users adding new dependencies? I'm worried about this workflow having more image rebuilds than users "need". I guess that's also a consideration w/r/t the `pyproject.toml`, if users do something like bump the version of their project (without changing any dependencies) it will force a rebuild of the modal Image.

J/w if we can be creative about trying to avoid such issues, since we're explicitly limiting this operation to _dependency_ provisioning.

Reply:
Spoke with @elias about this a bit and we were both still a little nervous that it won't be obvious that any changes to `pyproject.toml` will invalidate the whole Image. OTOH it's not obvious to me that there's a better solution.